tooling/ast_fuzzer/src/program/func.rs

```diff
@@ -85,6 +77,43 @@ pub(crate) fn hir_param(
     (pat, typ, vis)
 }
 
+/// Help avoid infinite recursion by limiting which function can call which other one.
+pub(super) fn can_call(
+    caller_id: FuncId,
+    caller_unconstrained: bool,
+    callee_id: FuncId,
+    callee_unconstrained: bool,
+) -> bool {
+    // Nobody should call `main`.
+    if callee_id == Program::main_id() {
```

Comment:
```suggestion
    if callee_id == Program::main_id() && !callee_unconstrained {
```

Reply:
Created a follow-up issue https://github.com/noir-lang/noir/issues/8549

This story is complicated enough without the extra source of potential errors 🙏